### PR TITLE
Remove encodings, take two

### PR DESCRIPTION
**Pull request by ralphtheninja**

### CODE REVIEW

File: lib/levelup.js [hunk 1]
```diff
@@ -19,17 +18,16 @@ var WriteError = errors.WriteError
...
 var promisify = util.promisify
 
-function getCallback (options, callback) {
-  return typeof options === 'function' ? options : callback
-}
+// TODO
+// * options are no longer needed for levelup itself, they are passed on to
+//   the underlying store during open phase, update docs!
+// * options used for .get, .put, .del and .batch are also passed on to
+//   the underlying store, update docs!
```
Comment (ralphtheninja) [inline lib/levelup.js]: removed in the next commit

File: lib/levelup.js [hunk 2]
```diff
@@ -286,20 +264,9 @@ LevelUP.prototype.batch = function (arr_, options, callback) {
...
-  options = extend({keys: true, values: true}, this.options, options)
-
-  options.keyEncoding = options.keyEncoding
-  options.valueEncoding = options.valueEncoding
-
-  options = this._codec.encodeLtgt(options)
-  options.keyAsBuffer = this._codec.keyAsBuffer(options)
-  options.valueAsBuffer = this._codec.valueAsBuffer(options)
-
+  options = extend({ keys: true, values: true }, options)
```
Comment (ralphtheninja) [inline lib/levelup.js]: Note. No use extending `this.options` anymore.

File: lib/levelup.js [hunk 3]
```diff
@@ -316,5 +283,20 @@ LevelUP.prototype.toString = function () {
   return 'LevelUP'
 }
 
+function getCallback (options, callback) {
+  return typeof options === 'function' ? options : callback
+}
+
+function getOptions (options) {
+  return typeof options === 'object' && options !== null ? options : {}
+}
```
Comment (ralphtheninja) [inline lib/levelup.js]: Moved this from `lib/util.js`. Maybe we can get rid of `lib/util.js` completely now.

File: lib/levelup.js [hunk 4]
```diff
@@ -145,62 +135,43 @@ LevelUP.prototype.isClosed = function () {
...
     }
-    if (callback) {
-      try {
-        value = self._codec.decodeValue(value, options)
-      } catch (e) {
-        return callback(new EncodingError(e))
-      }
-      callback(null, value)
-    }
+    if (callback) { callback(null, value) }
```
Comment (juliangruber) [inline lib/levelup.js]: we always have a callback now :)

File: lib/levelup.js [hunk 5]
```diff
@@ -210,25 +181,24 @@ LevelUP.prototype.put = function (key_, value_, options, callback) {
...
-  key = this._codec.encodeKey(key_, options)
-  value = this._codec.encodeValue(value_, options)
 
   this.db.put(key, value, options, function (err) {
     if (err) {
       return dispatchError(self, new WriteError(err), callback)
     }
-    self.emit('put', key_, value_)
+    self.emit('put', key, value)
     if (callback) { callback() }
```
Comment (juliangruber) [inline lib/levelup.js]: we always have a callback here too (should have done this in the promisify branch)

File: lib/levelup.js [hunk 6]
```diff
@@ -238,26 +208,28 @@ LevelUP.prototype.del = function (key_, options, callback) {
...
   options = getOptions(options)
-  key = this._codec.encodeKey(key_, options)
 
   this.db.del(key, options, function (err) {
     if (err) {
       return dispatchError(self, new WriteError(err), callback)
     }
-    self.emit('del', key_)
+    self.emit('del', key)
     if (callback) { callback() }
```
Comment (juliangruber) [inline lib/levelup.js]: as above

File: lib/levelup.js [hunk 7]
```diff
@@ -277,7 +249,7 @@ LevelUP.prototype.batch = function (arr_, options, callback) {
     if (err) {
       return dispatchError(self, new WriteError(err), callback)
     }
-    self.emit('batch', arr_)
+    self.emit('batch', arr)
     if (callback) { callback() }
```
Comment (juliangruber) [inline lib/levelup.js]: as above

File: lib/levelup.js (hunk 3 above)
Comment (juliangruber) [reply lib/levelup.js]: yeah, the promisify thing can go in npm :)

File: lib/levelup.js (hunk 4 above)
Comment (ralphtheninja) [reply lib/levelup.js]: Aah right, of course, duh! :)

Comment (ralphtheninja) [reply lib/levelup.js]: @juliangruber Noticed you missed that for `.open()` and `.close()`, fixing that as well.

File: lib/batch.js [hunk 8]
```diff
@@ -64,7 +63,7 @@ Batch.prototype.write = function (callback) {
 
   try {
     this.batch.write(function (err) {
-      if (err) { return dispatchError(levelup, new WriteError(err), callback) }
```
Comment (ralphtheninja) [inline lib/batch.js]: Moar complexity gone! :man_dancing: :dancer:

File: lib/levelup.js (hunk 4 above)
Comment (ralphtheninja) [reply lib/levelup.js]: Done.

File: lib/levelup.js (hunk 5 above)
Comment (ralphtheninja) [reply lib/levelup.js]: Done.

File: lib/levelup.js (hunk 6 above)
Comment (ralphtheninja) [reply lib/levelup.js]: Done.

File: lib/levelup.js (hunk 7 above)
Comment (ralphtheninja) [reply lib/levelup.js]: Done.

**Pull request merged**